src/main.rs

```diff
@@ -40,6 +42,13 @@ fn determine_state_positions<D: DFA>(
                 state_vec.push(next_state);
             }
         }
+        current_index += 1;
+    }
+
+    state_vec.shuffle(&mut thread_rng());
+    current_index = 0;
+
+    while let Some(&current_state) = state_vec.get(current_index) {
```

Comment:
Might as well drop the index in the second loop (since the vector is already built) and do `for &current_state in state_vec {`